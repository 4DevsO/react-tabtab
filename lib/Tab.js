'use strict';

var React = require('react');
var classNames = require('classnames');
var Tappable = require('react-tappable');

var Tab = React.createClass({
  displayName: 'Tab',

  handleTabDeleteButton: function handleTabDeleteButton() {
    this.props.handleTabDeleteButton(this.props.tabKey);
  },

  clickTab: function clickTab() {
    this.props.handleTabClick(this.props.tabKey);
  },

  render: function render() {
    var tabClass;
    var tabTmpl;

    if (this.props.status === 'active') {
      tabClass = classNames(this.props.style + 'tab', 'active');
    } else {
      tabClass = classNames(this.props.style + 'tab');
    }
    if (this.props.tabKey === 'ADD') {
      tabClass = classNames(this.props.style + 'tab', 'add');
    }

    // only show the delete button when it's active
    var closeButtonStyle;
    if (this.props.tabDeleteButton && this.props.status === "active") {
      closeButtonStyle = { display: 'inline-block' };
    } else {
      closeButtonStyle = { display: 'none' };
    }

    return React.createElement(
      Tappable,
      { onClick: this.clickTab },
      tabTmpl,
      React.createElement(
        'li',
        { className: tabClass },
        this.props.title,
        React.createElement(
          'div',
          { className: 'tabtab__folder__circCont', style: closeButtonStyle },
          React.createElement('div', { className: 'tabtab__folder__circle', onClick: this.handleTabDeleteButton })
        )
      )
    );
  }
});

module.exports = Tab;