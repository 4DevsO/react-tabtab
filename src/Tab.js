import React from 'react';
import styled from 'styled-components';

const TabStyle = styled.div`
  display: inline-block;
  color: ${props => props.active ? '#007bff' : 'black'};
  border-bottom: ${props => props.active ? '2px solid #007bff' : '1px solid #eee'};
  padding: 5px 10px;
  transition: color .3s cubic-bezier(.645, .045, .355, 1);
  &:hover {
    cursor: pointer;
    color: #007bff;
  }
`;

export default class Tab extends React.Component {
  constructor(props) {
    super(props);
    this.clickTab = this.clickTab.bind(this);
  }

  clickTab() {
    const {handleActiveIndex, index} = this.props;
    console.log(index)
    handleActiveIndex(index);
  }

  render() {
    const {customStyle, active} = this.props;
    // console.log(this.props)
    const TabComponent = customStyle || TabStyle;
    return (
      <TabComponent onClick={this.clickTab} active={active}>
        {this.props.children}
      </TabComponent>
    )
  }
}

export {
  TabStyle
};