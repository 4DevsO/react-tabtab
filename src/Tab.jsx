import React from 'react';
import ReactDOM from 'react-dom';
import classNames from 'classnames';
import Tappable from 'react-tappable';
import ItemTypes from './ItemTypes';
import { DragSource, DropTarget } from 'react-dnd';

const tabSource = {
  beginDrag: function(props) {
    props.beginDrag();
    return {
      tabKey: props.tabKey
    }
  }
}

// reference from https://github.com/gaearon/react-dnd/tree/master/examples/04%20Sortable/Simple
const tabTarget = {
  hover(props, monitor, component) {
    const dragIndex = monitor.getItem().tabKey;
    const hoverIndex = props.tabKey;

    // Don't replace items with themselves
    if (dragIndex === hoverIndex) {
      return;
    }

    // Determine rectangle on screen
    const hoverBoundingRect = ReactDOM.findDOMNode(component).getBoundingClientRect();

    // Get horizontal middle
    const hoverMiddleX = (hoverBoundingRect.right - hoverBoundingRect.left) / 2;

    // Determine mouse position
    const clientOffset = monitor.getClientOffset();

    const hoverClientX = hoverBoundingRect.right - clientOffset.x;

    // Dragging left
    if (dragIndex < hoverIndex && hoverClientX < hoverMiddleX) {
      return;
    }

    // Dragging right
    if (dragIndex > hoverIndex && hoverClientX > hoverMiddleX) {
      return;
    }

    props.moveTab(dragIndex, hoverIndex);
    // change the dragging tab index
    monitor.getItem().tabKey = hoverIndex;
  }
};

@DropTarget(ItemTypes.TAB, tabTarget, connect => ({
  connectDropTarget: connect.dropTarget()
}))
@DragSource(ItemTypes.TAB, tabSource, (connect, monitor) => ({
  connectDragSource: connect.dragSource(),
  isDragging: monitor.isDragging()
}))

export default class Tab extends React.Component {
  constructor(props) {
    super(props);
    this.handleTabDeleteButton = this.handleTabDeleteButton.bind(this);
    this.clickTab = this.clickTab.bind(this);
  }

  handleTabDeleteButton(e) {
    e.stopPropagation(); //prevent trigger clickTab function
    this.props.handleTabDeleteButton(this.props.tabKey);
  }

  clickTab() {
    this.props.handleTabClick(this.props.tabKey);
  }

  render() {
    let tabClass,
        closeButtonStyle;
    const {connectDragSource, connectDropTarget } = this.props;
    if (this.props.status === 'active') {
      tabClass = classNames(this.props.style + 'tab', 'active');
    } else {
      tabClass = classNames(this.props.style + 'tab');
    }

    // only show the delete button when it's active
    if (this.props.tabDeleteButton && this.props.status === "active") {
      closeButtonStyle = {display: 'inline-block'};
    } else {
      closeButtonStyle = {display: 'none'};
    }


    return connectDragSource(connectDropTarget(
      <span>
        <li className={tabClass} onClick={this.clickTab}>
          {this.props.title}
          <div className="tabtab__folder__circCont" style={closeButtonStyle}>
            <div className="tabtab__folder__circle" onClick={this.handleTabDeleteButton}></div>
          </div>
        </li>
      </span>
    ));
  }
}