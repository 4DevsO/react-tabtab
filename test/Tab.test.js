import React from 'react';
import Tab, { TabStyle } from '../src/Tab';
import { shallow } from 'enzyme';
import 'jest-styled-components'

test('render TabList', () => {
  const component = shallow(
    <Tab>
      <span>hi</span>
    </Tab>
  );
  expect(component).toMatchSnapshot();
})

test('custom ListStyle style', () => {
  const CustomTabStyle = TabStyle.extend`
    background-color: red;
  `;
  const component = shallow(
    <Tab customStyle={CustomTabStyle}>
      <span>hi</span>
    </Tab>
  );
  expect(component).toHaveStyleRule('background-color', 'red');  
})
