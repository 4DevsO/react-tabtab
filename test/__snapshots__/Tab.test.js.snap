// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render TabList 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Tab>
    <span>
      hi
    </span>
  </Tab>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <span>
        hi
      </span>,
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "hi",
      },
      "ref": null,
      "rendered": "hi",
      "type": "span",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <span>
          hi
        </span>,
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "hi",
        },
        "ref": null,
        "rendered": "hi",
        "type": "span",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
